test/suites/clustering_move.sh

```diff
@@ -0,0 +1,199 @@
+
+        return
+EOF
+
+  LXD_DIR="${LXD_ONE_DIR}" lxc move c1 --target @foobar3
+  LXD_DIR="${LXD_ONE_DIR}" lxc info c1 | grep -q "Location: node3"
+  LXD_DIR="${LXD_ONE_DIR}" lxc move c2 --target @foobar3
+  LXD_DIR="${LXD_ONE_DIR}" lxc info c2 | grep -q "Location: node3"
+
+  # Ensure that setting an invalid target won't interrupt the move and fall back to the built in behavior.
+  # Equally distribute the instances beforehand so that node1 will get selected.
+  LXD_DIR="${LXD_ONE_DIR}" lxc move c2 --target node2
+
+  cat << EOF | LXD_DIR="${LXD_ONE_DIR}" lxc config set instances.placement.scriptlet=-
+def instance_placement(request, candidate_members):
+        # Set invalid member target.
+        result = set_target("foo")
+        log_warn("Setting invalid member target result: ", result)
+
+        return
+EOF
+
+  LXD_DIR="${LXD_ONE_DIR}" lxc move c1 --target @foobar1
+  LXD_DIR="${LXD_ONE_DIR}" lxc info c1 | grep -q "Location: node1"
+
+  # If the scriptlet produces a runtime error, the move fails.
+  cat << EOF | LXD_DIR="${LXD_ONE_DIR}" lxc config set instances.placement.scriptlet=-
+def instance_placement(request, candidate_members):
+        # Try to access an invalid index (non existing member)
+        log_info("Accessing invalid field ", candidate_members[42])
+
+        return
+EOF
+
+  ! LXD_DIR="${LXD_ONE_DIR}" lxc move c1 --target @foobar2 || false
+
+  # If the scriptlet intentionally runs into an error, the move fails.
+  cat << EOF | LXD_DIR="${LXD_ONE_DIR}" lxc config set instances.placement.scriptlet=-
+def instance_placement(request, candidate_members):
+        log_error("instance placement not allowed") # Log placement error.
+
+        fail("Instance not allowed") # Fail to prevent instance creation.
+EOF
+
+  ! LXD_DIR="${LXD_ONE_DIR}" lxc move c1 --target @foobar2 || false
+
+  # Cleanup
+  LXD_DIR="${LXD_ONE_DIR}" lxc config unset instances.placement.scriptlet
+
+  # Perform project restriction tests.
+  # At this stage we have:
+  # - node1 in group foobar1,default
+  # - node2 in group foobar2,default
+  # - node3 in group foobar3,default
+  # - c1 is deployed on node1
+  # - c2 is deployed on node2
+  # - c3 is deployed on node3
+  # - default project restricted to cluster groups foobar1,foobar2
+  LXD_DIR="${LXD_ONE_DIR}" lxc project set default restricted=true
+  LXD_DIR="${LXD_ONE_DIR}" lxc project set default restricted.cluster.groups=foobar1,foobar2
+
+  # Moving to a node that is not a member of foobar1 or foobar2 will fail.
+  # The same applies for an unlisted group
+  ! LXD_DIR="${LXD_ONE_DIR}" lxc move c1 --target @foobar3 || false
+  ! LXD_DIR="${LXD_ONE_DIR}" lxc move c2 --target node3 || false
+
+  # Moving instances in between the restricted groups
+  LXD_DIR="${LXD_ONE_DIR}" lxc move c1 --target node2
+  LXD_DIR="${LXD_ONE_DIR}" lxc move c2 --target @foobar1
+  LXD_DIR="${LXD_ONE_DIR}" lxc move c3 --target node1
+
+  # Cleanup
+  LXD_DIR="${LXD_ONE_DIR}" lxc delete c1 c2 c3
```

Comment:
We tend to use the force flag (`-f`) when deleting instances.

Reply:
Added.